src/vs/base/browser/htmlContentRenderer.ts

```diff
@@ -20,9 +25,33 @@ export interface RenderOptions {
 	codeBlockRenderer?: (modeId: string, value: string) => string | TPromise<string>;
 }
 
-export function renderMarkedString(markedString: MarkedString, options: RenderOptions = {}): Node {
+export function renderMarkedString(modeService: IModeService, openerService: IOpenerService, markedString: MarkedString): Node {
 	const htmlContentElement = typeof markedString === 'string' ? { markdown: markedString } : { code: markedString };
-	return renderHtml(htmlContentElement, options);
+
+	return renderHtml(htmlContentElement, {
+		actionCallback: (content) => {
+			openerService.open(URI.parse(content)).then(void 0, onUnexpectedError);
+		},
+		codeBlockRenderer: (languageAlias, value): string | TPromise<string> => {
+			// According to the language server spec (https://github.com/Microsoft/language-server-protocol)
+			// A MarkedString is rendered as:
+			// - CASE 1: markdown if it is represented as a string
+			// - CASE 2: markdown code block of the given langauge if it is represented as a pair of a language and a value
+			//
+			// Because this function is only responsible for rendering code blocks, we only need to worry about handling
+			// CASE 2 here. We follow the spec exactly because in some cases (e.g. PHP), we want the content to be
+			// rendered using less strict rules than the official grammar (as we sometimes do in markdown.)
+			//
+			// Also note that we render the codeblock using TextMate grammar rather than using highlight.js as we do in the
+			// markdown preview. This is because we want renderings to be as consistent with the editor as possible.
+			value = '```' + languageAlias + '\n' + value + '\n ```';
+			const modeId = 'markdown';
```

Comment:
@mjbvz indeed, I did look into the other options we discussed, and I tried to enumerate the different options in my [comment here](https://github.com/Microsoft/vscode/pull/16503#issuecomment-265023706). Is there another option that you had in mind? 

Also, it never gets rendered as 
~~~markdown
```markdown
```php
print('codes')
```
```
~~~

In no place does 'markdown' get prepended by three backticks. Or is that not what you meant?